app/src/main/java/io/apicurio/registry/ccompat/store/RegistryStorageFacadeImpl.java

```diff
@@ -46,84 +53,144 @@
+                .stream()
+                .map(FacadeConverter::convertUnsigned)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public Schema findSchemaWithSubject(String subject, boolean checkDeletedSchema, String schema) throws ArtifactNotFoundException, RegistryStorageException {
+    public Schema getSchema(String subject, SchemaContent schema) throws ArtifactNotFoundException, RegistryStorageException {
         // TODO -- handle deleted?
-        ArtifactMetaDataDto amd = storage.getArtifactMetaData(subject, ContentHandle.create(schema));
+        ArtifactMetaDataDto amd = storage.getArtifactMetaData(subject, ContentHandle.create(schema.getSchema()));
         StoredArtifact storedArtifact = storage.getArtifactVersion(subject, amd.getVersion());
-        return toSchema(subject, storedArtifact);
+        return FacadeConverter.convert(subject, storedArtifact);
+    }
+
+    @Override
+    public CompletionStage<Long> createSchema(String subject, String schema) throws ArtifactAlreadyExistsException, ArtifactNotFoundException, RegistryStorageException {
+
+        // TODO Should this creation and updating of an artifact be a different operation?
+        // TODO method that returns a completion stage should not throw an exception
+        CompletionStage<ArtifactMetaDataDto> artifactMeta = createOrUpdateArtifact(subject, schema);
+
+        return artifactMeta.thenApply(ArtifactMetaDataDto::getGlobalId);
+    }
+
+    @Override
+    public int deleteSchema(String subject, String versionString) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException {
+        return FacadeConverter.convertUnsigned(parseVersionString(subject, versionString, version -> {
+            storage.deleteArtifactVersion(subject, version);
+            return version;
+        }));
+    }
+
+    @Override
+    public void createOrUpdateArtifactRule(String subject, RuleType type, RuleConfigurationDto dto) {
+        if (!doesArtifactRuleExist(subject, RuleType.COMPATIBILITY)) {
+            storage.createArtifactRule(subject, RuleType.COMPATIBILITY, dto);
+        } else {
+            storage.updateArtifactRule(subject, RuleType.COMPATIBILITY, dto);
+        }
     }
 
     @Override
-    public CompletionStage<Long> registerSchema(String subject, Integer id, Integer version, String schema) throws ArtifactAlreadyExistsException, ArtifactNotFoundException, RegistryStorageException {
-        ArtifactMetaDataDto metadata = null;
+    public void createOrUpdateGlobalRule(RuleType type, RuleConfigurationDto dto) {
+        if (!doesGlobalRuleExist(RuleType.COMPATIBILITY)) {
+            storage.createGlobalRule(RuleType.COMPATIBILITY, dto);
+        } else {
+            storage.updateGlobalRule(RuleType.COMPATIBILITY, dto);
+        }
+    }
+
+    private CompletionStage<ArtifactMetaDataDto> createOrUpdateArtifact(String subject, String schema) {
+        if (!doesArtifactExist(subject)) {
```

Comment:
This should rather be to catch AlreadyExists on create, then to do previous check.
Since while you check for an existing artifact, and you get false, one can be created.
Highly unlikely, but still better to handle it "defensively", as that way there is no concurrency error.

Reply:
Even doing that we probably have a race condition in the Streams storage implementation when running more than one instance, right?

Reply:
I'm not sure it would be a race condition, since all artifacts with same artifactId are handled on the same node, in ordered (sync) fashion.
For create, if two requests would be concurrent, one would fail -- see StreamsRegistryStorage impl.
For update, it would also fail -- if someone would delete the artifacts during that async op.